### ISSUE TITLE
Concretizing Pattern Match API

### ISSUE THREAD

**Issue by danieldietrich** [javaslang/javaslang#1157]
[follow-up of #1087]

For the 2.0.0 release we will fixate the following pattern match API:

**1) We will add visitors for all Javaslang types.**

E.g. types are Option, List, Future, ... These visitors are safe because the patterns are exhaustive.

```java
list.match(
  (x, xs) -> ...,
  () -> ...
)
```

**TODO (not part of 2.0.0):** Open question how to provide visitors for 3rd party types.

**2) We provide pre-generated patterns for all Javaslang types.**

The type must have an arity <= 2, i.e. decompose to Tuple0, Tuple1 or Tuple2.

```java
import static javaslang.API.*;

Match(list).of(
  Case(List("1", $_), head -> ...),
  Case(List(), () -> ...),
)
```

**TODO:** Is Case of type `Function<T, Option<R>>` or `PartialFunction<T, R>`, where

```java
interface PartialFunction<T, R> extends Function<T, R> {
    boolean isApplicable(T obj);
}
```

**3) We provide syntactic sugar for all Javaslang types.**

```java
// equivalent to Match(list).of(cases)
list.match(
  Case(List("1", $_), name -> ...),
  Case(List(), () -> ...),
)
```

**4) We add match guards in form of predicates.**

```java
Match(obj).of(
  // Pattern*<List<T>, ...>, Predicate<? super List<T>>, Function<? super List<T>, ? extends R>
  Case(List($(), List.empty()), head -> head != null, head -> 1)
)
```

Also without patterns. This will be the **substitute for the removed Match API of Javaslang 1.x**.

```java
Match(obj).of(
  Case(Predicate<???> predicate, Function<???, ? extends R> f),
  Case(Predicate<???> predicate, Supplier<? extends R> f), // will be ambiguous I think
  Case(Predicate<???> predicate, R> f)
)
```

Having a set of `javaslang.Predicates` would make life much easier. We will add them (isType(Class), isIn(...), isNull(), isNotNull(), etc).

**4.1) Maybe we need a special Case for types**

It will not be possible to use Predicates to 'expand' the type and to have a meaningful function:

```java
// not possible
Match(exception).of(
    Case(isType(NumberFormatException.class), numberFormatException -> ...),
    Case(isType(ClassCastException.class), classCastException -> ...),
    Case($_, ...) // catch all
)
```

Therefore we probably need a special Case for types:

```java
Match(exception).of(
    Case(NumberFormatException.class, numberFormatException -> ...),
    Case(ClassCastException.class, classCastException -> ...),
    Case($_, ...) // catch all
)
```

Uses classes will have one limitation, generics are not supported. This leads to the next point, full-blown structural pattern matching of 3rd party types.

**5) NOT PART OF 2.0.0: We provide an annotation processor for 3rd party types.**

Match.of(localDate).cases(
    Case(LocalDate(2015, $_, $()), (yy, dd) -> ...),
    Case(LocalDate(...), (yy, mm, dd) -> ...)
)

---

As we see, beside the visitor-pattern-like match (see **1)**), the API is uniform: `Match(obj).of(Case...)`

**Comment by danieldietrich** [comment]
Progress:

`[ ]` match guards
`[ ]` visitor for all types
`[ ]` syntactic sugar for all types
`[ ]` ~~special case for types~~ define a set of predicates
`[ ]` annotation processor + pre-generated patterns

**Comment by danieldietrich** [comment]
We will skip the features 'visitor for all types' and 'syntactic sugar for all types'. These will be added later if needed. Now it is closed-shop time for 2.0.0 final.
